File: src/engine/replay.ts
import type { Action, SourceRef } from './actions';
import { digestShort } from './digest';
import { applyAction, createGame, type DeckList } from './engine';
import type { GameState } from './state';
import type { PlayerIdx, TargetRef } from './types';

/** How a target is written on the wire, flat so both engines parse it the same. */
export interface WireRef {
  kind: string;
  player: number;
  index: number;
}

export interface WireAction {
  type: string;
  handIndex?: number;
  index?: number;
  slot?: number;
  source?: WireRef;
  powerIndex?: number;
  target?: WireRef;
  targets?: WireRef[];
  /** Party-only enemy pick. Written only when set, so 2-player replays never carry it. */
  enemy?: number;
}

export interface ReplayDeck {
  name: string;
  leaderId: string;
  cards: string[];
}

export interface ReplayStep {
  actor: number;
  action: WireAction;
  /** Digest of the state after this action, for pinpointing divergence. */
  digest: string;
}

export interface Replay {
  format: number;
  label: string;
  seed: number;
  startingPlayer: number;
  decks: ReplayDeck[];
  setupDigest: string;
  steps: ReplayStep[];
  finalDigest: string;
  winner: number;
  winReason: string | null;
}

// --- wire conversion --------------------------------------------------------

export function refToWire(ref: TargetRef): WireRef {
  switch (ref.kind) {
    case 'summon':
      return { kind: 'summon', player: ref.player, index: ref.slot };
    case 'leader':
      return { kind: 'leader', player: ref.player, index: 0 };
    case 'color':
      return { kind: 'color', player: 0, index: 0 };
    default:
      return { kind: ref.kind, player: ref.player, index: ref.index };
  }
}

export function refFromWire(w: WireRef): TargetRef {
  const player = w.player as PlayerIdx;
  switch (w.kind) {
    case 'summon':
      return { kind: 'summon', player, slot: w.index };
    case 'leader':
      return { kind: 'leader', player };
    case 'hand':
      return { kind: 'hand', player, index: w.index };
    case 'supporter':
      return { kind: 'supporter', player, index: w.index };
    case 'debt':
      return { kind: 'debt', player, index: w.index };
    case 'discard':
      return { kind: 'discard', player, index: w.index };
    default:
      return { kind: 'color', color: 'P' };
  }
}

export function actionToWire(a: Action): WireAction {
  const enemy = 'enemy' in a && a.enemy !== undefined ? { enemy: a.enemy } : {};
  switch (a.type) {
    case 'PLAY_STAGE':
      return { type: a.type, handIndex: a.handIndex, ...enemy };
    case 'PLAY_SUPPORTER':
      return { type: a.type, handIndex: a.handIndex };
    case 'REPLACE_SUMMON':
      return {
        type: a.type,
        handIndex: a.handIndex,
        targets: (a.targets ?? []).map(refToWire),
        ...enemy,
      };
    case 'PAY_FLIP':
      return a.handIndex === undefined
        ? { type: a.type, ...enemy }
        : { type: a.type, handIndex: a.handIndex, ...enemy };
    case 'SAP_SUPPORTER':
      return { type: a.type, index: a.index };
    case 'PLAY_SUMMON':
      return {
        type: a.type,
        handIndex: a.handIndex,
        slot: a.slot,
        targets: (a.targets ?? []).map(refToWire),
        ...enemy,
      };
    case 'CAST_SPELL':
      return { type: a.type, handIndex: a.handIndex, targets: a.targets.map(refToWire), ...enemy };
    case 'CAST_TRAP':
      return { type: a.type, handIndex: a.handIndex, targets: a.targets.map(refToWire) };
    case 'ACTIVATE_POWER':
      return {
        type: a.type,
        source: refToWire(a.source),
        powerIndex: a.powerIndex,
        targets: a.targets.map(refToWire),
        ...enemy,
      };
    case 'DECLARE_ATTACK':
      return { type: a.type, source: refToWire(a.source), target: refToWire(a.target) };
    case 'RESOLVE_CHOICE':
      return {
        type: a.type,
        ...(a.pick ? { target: refToWire(a.pick) } : {}),
        ...(a.index !== undefined ? { index: a.index } : {}),
      };
    default:
      return { type: a.type };
  }
}

export function actionFromWire(w: WireAction): Action {
  const targets = (w.targets ?? []).map(refFromWire);
  const enemy = w.enemy !== undefined ? { enemy: w.enemy as PlayerIdx } : {};
  switch (w.type) {
    case 'PLAY_SUPPORTER':
      return { type: 'PLAY_SUPPORTER', handIndex: w.handIndex ?? 0 };
    case 'SAP_SUPPORTER':
      return { type: 'SAP_SUPPORTER', index: w.index ?? 0 };
    case 'PLAY_SUMMON':
      return {
        type: 'PLAY_SUMMON',
        handIndex: w.handIndex ?? 0,
        slot: w.slot ?? 0,
        targets,
        ...enemy,
      };
    case 'CAST_SPELL':
      return { type: 'CAST_SPELL', handIndex: w.handIndex ?? 0, targets, ...enemy };
    case 'PLAY_STAGE':
      return { type: 'PLAY_STAGE', handIndex: w.handIndex ?? 0, ...enemy };
    case 'ACTIVATE_POWER':
      return {
        type: 'ACTIVATE_POWER',
        source: refFromWire(w.source!) as SourceRef,
        powerIndex: w.powerIndex ?? 0,
        targets,
        ...enemy,
      };
    case 'DECLARE_ATTACK':
      return {
        type: 'DECLARE_ATTACK',
        source: refFromWire(w.source!) as SourceRef,
        target: refFromWire(w.target!),
      };
    case 'CAST_TRAP':
      return { type: 'CAST_TRAP', handIndex: w.handIndex ?? 0, targets };
    case 'PASS_RESPONSE':
      return { type: 'PASS_RESPONSE' };
    case 'RESOLVE_CHOICE':
      return {
        type: 'RESOLVE_CHOICE',
        ...(w.target ? { pick: refFromWire(w.target) } : {}),
        ...(w.index !== undefined ? { index: w.index } : {}),
      };
    case 'REPLACE_SUMMON':
      return { type: 'REPLACE_SUMMON', handIndex: w.handIndex ?? 0, targets, ...enemy };
    case 'DECLINE_REPLACE':
      return { type: 'DECLINE_REPLACE' };
    case 'PAY_FLIP':
      return w.handIndex === undefined
        ? { type: 'PAY_FLIP', ...enemy }
        : { type: 'PAY_FLIP', handIndex: w.handIndex, ...enemy };
    case 'DECLINE_FLIP':
      return { type: 'DECLINE_FLIP' };
    case 'END_TURN':
      return { type: 'END_TURN' };
    case 'CONCEDE':
      return { type: 'CONCEDE' };
    default:
      throw new Error(`unknown action type ${w.type}`);
  }
}

// --- recording and verifying ------------------------------------------------

export interface Recorder {
  state: GameState;
  replay: Replay;
}

export function startRecording(
  decks: [DeckList, DeckList],
  seed: number,
  startingPlayer: PlayerIdx,
  label: string,
): Recorder {
  const state = createGame(decks, seed, startingPlayer);
  return {
    state,
    replay: {
      format: 1,
      label,
      seed,
      startingPlayer,
      decks: decks.map((d) => ({ name: d.name, leaderId: d.leaderId, cards: [...d.cards] })),
      setupDigest: digestShort(state),
      steps: [],
      finalDigest: '',
      winner: -1,
      winReason: null,
    },
  };
}

/** Applies an action and appends it to the recording. Throws if it was illegal. */
export function recordAction(rec: Recorder, actor: PlayerIdx, action: Action): void {
  const res = applyAction(rec.state, actor, action);
  if (!res.ok) throw new Error(`recorded an illegal ${action.type}: ${res.error}`);
  rec.state = res.state;
  rec.replay.steps.push({
    actor,
    action: actionToWire(action),
    digest: digestShort(rec.state),
  });
}

export function finishRecording(rec: Recorder): Replay {
  rec.replay.finalDigest = digestShort(rec.state);
  rec.replay.winner = rec.state.winner === null ? -1 : rec.state.winner;
  rec.replay.winReason = rec.state.winReason;
  return rec.replay;
}

export interface ReplayResult {
  ok: boolean;
  stepIndex: number;
  detail: string | null;
}

/**
 * Re-runs a replay and reports the first step that does not match. This is what
 * makes the two engines checkable against each other and what makes a rules
 * change visible as an exact list of games whose outcome moved.
 */
export function verifyReplay(replay: Replay): ReplayResult {
  if (replay.decks.length !== 2) {
    return { ok: false, stepIndex: -1, detail: 'replay needs exactly two decks' };
  }
  const decks: [DeckList, DeckList] = [
    { name: replay.decks[0].name, leaderId: replay.decks[0].leaderId, cards: replay.decks[0].cards },
    { name: replay.decks[1].name, leaderId: replay.decks[1].leaderId, cards: replay.decks[1].cards },
  ];
  let state = createGame(decks, replay.seed, replay.startingPlayer as PlayerIdx);

  if (replay.setupDigest) {
    const got = digestShort(state);
    if (got !== replay.setupDigest) {
      return { ok: false, stepIndex: -1, detail: `setup digest ${got} != ${replay.setupDigest}` };
    }
  }

  for (let i = 0; i < replay.steps.length; i++) {
    const step = replay.steps[i];
    const action = actionFromWire(step.action);
    const res = applyAction(state, step.actor as PlayerIdx, action);
    if (!res.ok) {
      return { ok: false, stepIndex: i, detail: `${action.type} rejected: ${res.error}` };
    }
    state = res.state;
    if (step.digest) {
      const got = digestShort(state);
      if (got !== step.digest) {
        return {
          ok: false,
          stepIndex: i,
          detail: `after ${action.type}: digest ${got} != ${step.digest}`,
        };
      }
    }
  }

  if (replay.finalDigest) {
    const got = digestShort(state);
    if (got !== replay.finalDigest) {
      return {
        ok: false,
        stepIndex: replay.steps.length,
        detail: `final digest ${got} != ${replay.finalDigest}`,
      };
    }
  }
  const winner = state.winner === null ? -1 : state.winner;
  if (winner !== replay.winner) {
    return {
      ok: false,
      stepIndex: replay.steps.length,
      detail: `winner ${winner} != ${replay.winner}`,
    };
  }
  return { ok: true, stepIndex: -1, detail: null };
}

File: csharp/Selatza.Engine/Replay.cs
using System.Text;
using System.Text.Json;
using System.Text.Json.Serialization;

namespace Selatza;

public sealed class ReplayDeck
{
    [JsonPropertyName("name")] public string Name { get; set; } = "";
    [JsonPropertyName("leaderId")] public string LeaderId { get; set; } = "";
    [JsonPropertyName("cards")] public List<string> Cards { get; set; } = new();
}

public sealed class ReplayStep
{
    [JsonPropertyName("actor")] public int Actor { get; set; }
    [JsonPropertyName("action")] public JsonElement Action { get; set; }
    /// <summary>Digest of the state after this action, for pinpointing divergence.</summary>
    [JsonPropertyName("digest")] public string Digest { get; set; } = "";
}

public sealed class Replay
{
    [JsonPropertyName("format")] public int Format { get; set; } = 1;
    [JsonPropertyName("label")] public string Label { get; set; } = "";
    [JsonPropertyName("seed")] public int Seed { get; set; }
    [JsonPropertyName("startingPlayer")] public int StartingPlayer { get; set; }
    [JsonPropertyName("decks")] public List<ReplayDeck> Decks { get; set; } = new();
    [JsonPropertyName("setupDigest")] public string SetupDigest { get; set; } = "";
    [JsonPropertyName("steps")] public List<ReplayStep> Steps { get; set; } = new();
    [JsonPropertyName("finalDigest")] public string FinalDigest { get; set; } = "";
    [JsonPropertyName("winner")] public int Winner { get; set; } = -1;
    [JsonPropertyName("winReason")] public string? WinReason { get; set; }

    public static readonly JsonSerializerOptions JsonOpts = new()
    {
        WriteIndented = false,
        DefaultIgnoreCondition = JsonIgnoreCondition.Never,
    };

    public static Replay Load(string path) =>
        JsonSerializer.Deserialize<Replay>(File.ReadAllText(path), JsonOpts)
        ?? throw new InvalidDataException($"could not parse replay {path}");

    public void Save(string path) =>
        File.WriteAllText(path, JsonSerializer.Serialize(this, JsonOpts));
}

public readonly record struct ReplayResult(bool Ok, int StepIndex, string? Detail)
{
    public static ReplayResult Good => new(true, -1, null);
}

/// <summary>
/// Replays are the contract between the two engines and between versions of one
/// engine. A replay stores only a seed, the deck lists and the actions taken, so
/// re-running it rebuilds the whole game; the stored digests say where any
/// divergence started rather than just that one happened.
/// </summary>
public static class Replays
{
    public static GameAction ParseAction(JsonElement e)
    {
        var type = e.GetProperty("type").GetString() ?? "";
        var kind = ParseType(type);
        bool hasHand = e.TryGetProperty("handIndex", out _);
        var a = new GameAction { Type = kind };
        return a with
        {
            HandIndex = kind == ActionType.PayFlip && !hasHand ? -1 : GetInt(e, "handIndex"),
            Index = GetInt(e, "index"),
            Slot = GetInt(e, "slot"),
            PowerIndex = GetInt(e, "powerIndex"),
            Source = GetRef(e, "source") ?? default,
            Target = GetRef(e, "target") ?? default,
            Targets = GetRefs(e, "targets"),
            Pick = kind == ActionType.ResolveChoice ? GetRef(e, "target") : null,
            ChoiceIndex = kind == ActionType.ResolveChoice && e.TryGetProperty("index", out var ciEl)
                && ciEl.ValueKind == System.Text.Json.JsonValueKind.Number
                ? ciEl.GetInt32()
                : null,
        };
    }

    public static string ActionToJson(GameAction a)
    {
        var sb = new StringBuilder();
        sb.Append("{\"type\":\"").Append(TypeName(a.Type)).Append('"');
        switch (a.Type)
        {
            case ActionType.PlaySupporter:
                sb.Append(",\"handIndex\":").Append(a.HandIndex);
                break;
            case ActionType.ReplaceSummon:
                sb.Append(",\"handIndex\":").Append(a.HandIndex)
                  .Append(",\"targets\":").Append(RefsToJson(a.Targets));
                break;
            case ActionType.SapSupporter:
                sb.Append(",\"index\":").Append(a.Index);
                break;
            case ActionType.PlaySummon:
                sb.Append(",\"handIndex\":").Append(a.HandIndex).Append(",\"slot\":").Append(a.Slot)
                  .Append(",\"targets\":").Append(RefsToJson(a.Targets));
                break;
            case ActionType.PlayStage:
                sb.Append(",\"handIndex\":").Append(a.HandIndex);
                break;
            case ActionType.PayFlip:
                if (a.HandIndex >= 0) sb.Append(",\"handIndex\":").Append(a.HandIndex);
                break;
            case ActionType.CastSpell:
            case ActionType.CastTrap:
                sb.Append(",\"handIndex\":").Append(a.HandIndex)
                  .Append(",\"targets\":").Append(RefsToJson(a.Targets));
                break;
            case ActionType.ActivatePower:
                sb.Append(",\"source\":").Append(RefToJson(a.Source))
                  .Append(",\"powerIndex\":").Append(a.PowerIndex)
                  .Append(",\"targets\":").Append(RefsToJson(a.Targets));
                break;
            case ActionType.DeclareAttack:
                sb.Append(",\"source\":").Append(RefToJson(a.Source))
                  .Append(",\"target\":").Append(RefToJson(a.Target));
                break;
            case ActionType.ResolveChoice:
                if (a.Pick is { } pick) sb.Append(",\"target\":").Append(RefToJson(pick));
                if (a.ChoiceIndex is { } ci) sb.Append(",\"index\":").Append(ci);
                break;
        }
        sb.Append('}');
        return sb.ToString();
    }

    public static string RefToJson(TargetRef r) =>
        $"{{\"kind\":\"{KindName(r.Kind)}\",\"player\":{r.Player},\"index\":{r.Index}}}";

    public static string RefsToJson(TargetRef[] refs) =>
        "[" + string.Join(",", refs.Select(RefToJson)) + "]";

    /// <summary>Re-runs a replay and reports the first step that does not match.</summary>
    public static ReplayResult Verify(Replay replay)
    {
        var decks = replay.Decks;
        if (decks.Count != 2) return new ReplayResult(false, -1, "replay needs exactly two decks");
        var state = Engine.CreateGame(
            new DeckList { Name = decks[0].Name, LeaderId = decks[0].LeaderId, Cards = decks[0].Cards },
            new DeckList { Name = decks[1].Name, LeaderId = decks[1].LeaderId, Cards = decks[1].Cards },
            replay.Seed,
            replay.StartingPlayer);

        if (replay.SetupDigest.Length > 0)
        {
            var got = Digest.Short(state);
            if (got != replay.SetupDigest)
            {
                return new ReplayResult(false, -1, $"setup digest {got} != {replay.SetupDigest}");
            }
        }

        for (int i = 0; i < replay.Steps.Count; i++)
        {
            var step = replay.Steps[i];
            var action = ParseAction(step.Action);
            var res = Engine.Apply(state, step.Actor, action);
            if (!res.Ok) return new ReplayResult(false, i, $"{action.Type} rejected: {res.Error}");
            state = res.State!;
            if (step.Digest.Length > 0)
            {
                var got = Digest.Short(state);
                if (got != step.Digest)
                {
                    return new ReplayResult(false, i,
                        $"after {action.Type}: digest {got} != {step.Digest}");
                }
            }
        }

        if (replay.FinalDigest.Length > 0)
        {
            var got = Digest.Short(state);
            if (got != replay.FinalDigest)
            {
                return new ReplayResult(false, replay.Steps.Count,
                    $"final digest {got} != {replay.FinalDigest}");
            }
        }
        if (replay.Winner != state.Winner)
        {
            return new ReplayResult(false, replay.Steps.Count,
                $"winner {state.Winner} != {replay.Winner}");
        }
        return ReplayResult.Good;
    }

    private static int GetInt(JsonElement e, string name) =>
        e.TryGetProperty(name, out var v) && v.ValueKind == JsonValueKind.Number ? v.GetInt32() : 0;

    private static TargetRef? GetRef(JsonElement e, string name)
    {
        if (!e.TryGetProperty(name, out var v) || v.ValueKind != JsonValueKind.Object) return null;
        return ParseRef(v);
    }

    private static TargetRef[] GetRefs(JsonElement e, string name)
    {
        if (!e.TryGetProperty(name, out var v) || v.ValueKind != JsonValueKind.Array)
        {
            return Array.Empty<TargetRef>();
        }
        var list = new List<TargetRef>();
        foreach (var item in v.EnumerateArray()) list.Add(ParseRef(item));
        return list.ToArray();
    }

    private static TargetRef ParseRef(JsonElement v)
    {
        var kind = v.GetProperty("kind").GetString() ?? "summon";
        int player = v.TryGetProperty("player", out var p) ? p.GetInt32() : 0;
        int index = v.TryGetProperty("index", out var i) ? i.GetInt32() :
            v.TryGetProperty("slot", out var s) ? s.GetInt32() : 0;
        return new TargetRef(ParseKind(kind), player, index);
    }

    public static TargetKind ParseKind(string k) => k switch
    {
        "summon" => TargetKind.Summon,
        "leader" => TargetKind.Leader,
        "hand" => TargetKind.Hand,
        "supporter" => TargetKind.Supporter,
        "debt" => TargetKind.Debt,
        "discard" => TargetKind.Discard,
        _ => TargetKind.ColorPick,
    };

    public static string KindName(TargetKind k) => k switch
    {
        TargetKind.Summon => "summon",
        TargetKind.Leader => "leader",
        TargetKind.Hand => "hand",
        TargetKind.Supporter => "supporter",
        TargetKind.Debt => "debt",
        TargetKind.Discard => "discard",
        _ => "color",
    };

    public static ActionType ParseType(string t) => t switch
    {
        "PLAY_SUPPORTER" => ActionType.PlaySupporter,
        "SAP_SUPPORTER" => ActionType.SapSupporter,
        "PLAY_SUMMON" => ActionType.PlaySummon,
        "CAST_SPELL" => ActionType.CastSpell,
        "PLAY_STAGE" => ActionType.PlayStage,
        "ACTIVATE_POWER" => ActionType.ActivatePower,
        "DECLARE_ATTACK" => ActionType.DeclareAttack,
        "CAST_TRAP" => ActionType.CastTrap,
        "PASS_RESPONSE" => ActionType.PassResponse,
        "RESOLVE_CHOICE" => ActionType.ResolveChoice,
        "REPLACE_SUMMON" => ActionType.ReplaceSummon,
        "DECLINE_REPLACE" => ActionType.DeclineReplace,
        "PAY_FLIP" => ActionType.PayFlip,
        "DECLINE_FLIP" => ActionType.DeclineFlip,
        "END_TURN" => ActionType.EndTurn,
        "CONCEDE" => ActionType.Concede,
        _ => throw new InvalidDataException($"unknown action type {t}"),
    };

    public static string TypeName(ActionType t) => t switch
    {
        ActionType.PlaySupporter => "PLAY_SUPPORTER",
        ActionType.SapSupporter => "SAP_SUPPORTER",
        ActionType.PlaySummon => "PLAY_SUMMON",
        ActionType.CastSpell => "CAST_SPELL",
        ActionType.PlayStage => "PLAY_STAGE",
        ActionType.ActivatePower => "ACTIVATE_POWER",
        ActionType.DeclareAttack => "DECLARE_ATTACK",
        ActionType.CastTrap => "CAST_TRAP",
        ActionType.PassResponse => "PASS_RESPONSE",
        ActionType.ResolveChoice => "RESOLVE_CHOICE",
        ActionType.ReplaceSummon => "REPLACE_SUMMON",
        ActionType.DeclineReplace => "DECLINE_REPLACE",
        ActionType.PayFlip => "PAY_FLIP",
        ActionType.DeclineFlip => "DECLINE_FLIP",
        ActionType.EndTurn => "END_TURN",
        _ => "CONCEDE",
    };
}
